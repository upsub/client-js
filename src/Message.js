export default class Message {
  /**
   * Create a new message instance
   * @param {Object} headers
   * @param {Mixed} payload
   */
  constructor (headers = {}, payload) {
    this.headers = headers
    this.payload = payload
  }

  /**
   * Get header upsub-message-type and return its value
   * @return {String}
   */
  get type () {
    return this.headers['upsub-message-type']
  }

  /**
   * Set header upsub-message-type
   * @param  {String} value
   */
  set type (value) {
    this.headers['upsub-message-type'] = value
  }

  /**
   * Encode message so its ready to be send
   * @return {String}
   */
  encode () {
    return JSON.stringify({
      headers: this.headers,
      payload: JSON.stringify(this.payload)
    })
  }

  /**
   * Decode message and create Message object from it
   * @param  {String} message
   * @return {Message}
   */
  static decode (message) {
    const { headers, payload } = JSON.parse(message)
    return new Message(headers, payload)
  }

  /**
   * Create a new bacth message
   * @param  {Message} messages which should be send together
   * @return {Message}
   */
  static batch (...messages) {
    const headers = {
      'upsub-message-type': Message.BATCH
    }

    return new Message(headers, messages)
  }

  /**
   * Create a new ping message
   * @return {Message}
   */
  static ping () {
    const headers = {
      'upsub-message-type': Message.PING
    }

    return new Message(headers)
  }

  /**
   * Create a new pong message
   * @return {Message}
   */
  static pong () {
    const headers = {
      'upsub-message-type': Message.PONG
    }

    return new Message(headers)
  }

  /**
   * Create a new text message
   * @param  {String} channel
   * @param  {Mixed} payload
   * @return {Message}
   */
  static text (channel, payload) {
    const headers = {
      'upsub-message-type': Message.TEXT,
      'upsub-channel': channel
    }
    return new Message(headers, payload)
  }

  /**
   * Create a new subscribe message
   * @param  {String} channels
   * @return {Message}
   */
  static subscribe (...channels) {
    const headers = {
      'upsub-message-type': Message.SUBSCRIBE,
    }
    return new Message(headers, channels.join(','))
  }

  /**
   * Create a new unsubscribe message
   * @param  {String} channels
   * @return {Message}
   */
  static unsubscribe (...channels) {
    const headers = {
      'upsub-message-type': Message.UNSUBSCRIBE
    }

    return new Message(headers, channels.join(','))
  }

  /**
   * Message type TEXT
   */
  static get TEXT () {
    return 'text'
  }

  /**
   * Message type BATCH
   */
  static get BATCH () {
    return 'batch'
  }

  /**
   * Message type SUBSCRIBE
   */
  static get SUBSCRIBE () {
    return 'subscribe'
  }

  /**
   * Message type UNSUBSCRIBE
   */
  static get UNSUBSCRIBE () {
    return 'unsubscribe'
  }

  /**
   * Message type PING
   */
  static get PING () {
    return 'ping'
  }

  /**
   * Message type PONG
   */
  static get PONG () {
    return 'pong'
  }

}
