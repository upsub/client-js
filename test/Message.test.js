/* eslint-env jest */
import Message from 'Message'

test('Should create a new message', () => {
  const msg = new Message({ 'header-key': 'value' }, { key: 'value' })
  expect(msg).toMatchSnapshot()
})

test('msg.type should return message type from header', () => {
  const msg = new Message({ 'upsub-message-type': 'text' })
  expect(msg.type).toBe('text')
})

test('Should set new message type through msg.type = "text"', () => {
  const msg = new Message()
  msg.type = 'batch'
  expect(msg.type).toBe('batch')
  expect(msg.headers).toEqual({ 'upsub-message-type': 'batch' })
})

test('Should encode message to string format', () => {
  const msg = new Message({ 'header-key': 'header-value' }, 'data')
  expect(msg.encode()).toMatchSnapshot()
  expect(msg.toString()).toEqual(msg.encode())
})

test('Should parse a string and create a new message', () => {
  const msg = Message.decode(
    JSON.stringify({
      headers: { 'upsub-message-type': 'text' },
      payload: JSON.stringify('data')
    })
  )

  expect(msg).toMatchSnapshot()
})

test('Should create a batch message', () => {
  const msg = Message.batch(new Message())
  expect(msg).toMatchSnapshot()
})

test('Should create a ping message', () => {
  const msg = Message.ping()
  expect(msg).toMatchSnapshot()
})

test('Should create a pong message', () => {
  const msg = Message.pong()
  expect(msg).toMatchSnapshot()
})

test('Should create a text message', () => {
  const msg = Message.text('channel', 'data')
  expect(msg).toMatchSnapshot()
})

test('Should create a subscribe message', () => {
  const msg = Message.subscribe('channel')
  const msg2 = Message.subscribe('channel', 'channel-2')
  expect(msg).toMatchSnapshot()
  expect(msg2).toMatchSnapshot()
})

test('Should create an unsubscribe message', () => {
  const msg = Message.unsubscribe('channel')
  const msg2 = Message.unsubscribe('channel', 'channel-2')
  expect(msg).toMatchSnapshot()
  expect(msg2).toMatchSnapshot()
})

test('Should export message types', () => {
  expect(Message.TEXT).toBe('text')
  expect(Message.BATCH).toBe('batch')
  expect(Message.SUBSCRIBE).toBe('subscribe')
  expect(Message.UNSUBSCRIBE).toBe('unsubscribe')
  expect(Message.PING).toBe('ping')
  expect(Message.PONG).toBe('pong')
})
