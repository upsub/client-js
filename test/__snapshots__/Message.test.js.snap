// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create a batch message 1`] = `
Message {
  "headers": Object {
    "upsub-message-type": "batch",
  },
  "payload": Array [
    Message {
      "headers": Object {},
      "payload": undefined,
    },
  ],
}
`;

exports[`Should create a new message 1`] = `
Message {
  "headers": Object {
    "header-key": "value",
  },
  "payload": Object {
    "key": "value",
  },
}
`;

exports[`Should create a ping message 1`] = `
Message {
  "headers": Object {
    "upsub-message-type": "ping",
  },
  "payload": undefined,
}
`;

exports[`Should create a pong message 1`] = `
Message {
  "headers": Object {
    "upsub-message-type": "pong",
  },
  "payload": undefined,
}
`;

exports[`Should create a subscribe message 1`] = `
Message {
  "headers": Object {
    "upsub-message-type": "subscribe",
  },
  "payload": "channel",
}
`;

exports[`Should create a subscribe message 2`] = `
Message {
  "headers": Object {
    "upsub-message-type": "subscribe",
  },
  "payload": "channel,channel-2",
}
`;

exports[`Should create a text message 1`] = `
Message {
  "headers": Object {
    "upsub-channel": "channel",
    "upsub-message-type": "text",
  },
  "payload": "data",
}
`;

exports[`Should create an unsubscribe message 1`] = `
Message {
  "headers": Object {
    "upsub-message-type": "unsubscribe",
  },
  "payload": "channel",
}
`;

exports[`Should create an unsubscribe message 2`] = `
Message {
  "headers": Object {
    "upsub-message-type": "unsubscribe",
  },
  "payload": "channel,channel-2",
}
`;

exports[`Should encode message to string format 1`] = `"{\\"headers\\":{\\"header-key\\":\\"header-value\\"},\\"payload\\":\\"\\\\\\"data\\\\\\"\\"}"`;

exports[`Should parse a string and create a new message 1`] = `
Message {
  "headers": Object {
    "upsub-message-type": "text",
  },
  "payload": "data",
}
`;
